Fix dispatch dict passed to `MergeStrategy` not getting updated for future fetch's

Marenz opened the pull request:
* We are also switching to use the same `now` while evaluating other dispatches in the MergeStrategy, but this is just an enhancement and not a fix
* We still used logging without "path" in some places, that was fixed

Marenz committed:
Use proper logging paths

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

Marenz committed:
Use the same `now` for all in merge check

Requires v0.11.1 of `frequenz-client-dispatch` for the new `started_at` method.

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

Marenz committed:
Reset release notes

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

Marenz committed:
Adjust logging to be more informative and clear

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

ela-kotulska-frequenz commented on src/frequenz/dispatch/_merge_strategies.py:
```diff
@@ -58,7 +58,8 @@ def filter(
         )
 
         _logger.debug(
-            "stop event %s because other_dispatches_running=%s",
+            "%s stop event %s because other_dispatches_running=%s",
+            "Ignoring" if other_dispatches_running else "Allowing",
             dispatch.id,
             other_dispatches_running,
         )
```
You could print IDs of running dispatches ( if there are any )

llucax commented on pyproject.toml:
```diff
@@ -40,7 +40,7 @@ dependencies = [
   "frequenz-sdk >= 1.0.0-rc2100, < 1.0.0-rc2200",
   "frequenz-channels >= 1.6.1, < 2.0.0",
-  "frequenz-client-dispatch >= 0.11.0, < 0.12.0",
+  "frequenz-client-dispatch >= 0.11.1, < 0.12.0",
```
This dependency update seems to be misplaced, or is it necessary to call `super().started_at()`? If that is the case please mention this in the commit message, other we need to just guess when reviewing.

Marenz replied:
Yeah, started_at was introduced in 0.11.1

llucax reviewed (commented):
As usual, the PR and commit title and description could use some improvement. It is hard to review when they say so little. Could you please please please put a bit more thought into those? Please :pray: ? For the PR description just copying the copilot review summary is a big improvement.

For example in this PR it is mostly irrelevant that you are using `started_at`, that tells me almost nothing. What is important is you are fixing a bug caused by not having a consistent `now` used in different path of the code to decide on how to merge dispatches.

Marenz committed:
Fix filter not seeing the latest dispatches after periodic fetch()'s

By using a workflow that avoids reassigning the dispatches dict.

When `fetch()`, which runs periodically, would assign
a fresh map to `self._dispatches` the filter would
not see the latest dispatches, as it would still
reference the old map.

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

Marenz committed:
Minor: Extend documentation phrasing

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

Marenz commented:
>  commit title and description 

tbh, I disagree. The commit titles and descriptions are all perfectly fine imo.

PR, yeah, okay, that grew a bit and needed updating.

I also don't think a required dependency update necessarily needs to be mentioned if it is so minor as here, and the fact that its part of that particular commit is in itself a documentation of its need, at least that was my perspective.... 

I don't really see how _any_ of the other commit titles & content is lacking anything to make them understandable.

Marenz commented:
* Updated to print all other active dispatch ids
* Updated to mention the need of the version update in the according PR description.
* Updated another commit description to be even more clear

Marenz commented on src/frequenz/dispatch/_merge_strategies.py:
```diff
@@ -39,24 +42,38 @@ def filter(
+        if other_dispatches_running:
+            _logger.debug(
+                "Active other dispatches: %s",
+                (dispatch.id for dispatch in running_dispatch_list),
```
oh damn.

llucax commented:
Without wanting to make a huge thing about this, I will just comment on this one, as I think it can bring some clarity:

> I also don't think a required dependency update necessarily needs to be mentioned if it is so minor as here, and the fact that its part of that particular commit is in itself a documentation of its need, at least that was my perspective....

I've seen many times people that sneak an unrelated change into a commit just because it was there. And I don't even think that's obscenely wrong, I can live with that when there is intention. The problem is when it was a mistake, and you were working on something else that needed the bump, but this PR doesn't really need it, and that dependency bump can have real world impact.

So what it is important for me is declaring intention. If the bump was needed by the commit, that's good to know, and good to know why, even if it could be inferred from the diff, just as an extra verification. If the bump was just done because one is lazy about putting it in a separate commit, **and the bump is harmless**, that should also be explicitly stated, so I can actually do a review, which is about detecting issues early.

This is my main point. I want to be able to tell between mistakes and intended changes.

llucax commented on src/frequenz/dispatch/_bg_service.py:
```diff
@@ -396,23 +396,30 @@ async def _fetch(self, timer: Timer) -> None:
+            if dispatch_id in new_dispatches:
+                continue
+
+            dispatch = self._dispatches.pop(dispatch_id)
```
`pop()` will raise if the key is not found, not sure if this could happen if you are dealing with some dict inconsistencies here. If it could happen you should either first check if it exists or catch the `KeyError` `pop` could raise.

Marenz replied:
`old_dispatches == keys of self._dispatches`, so it is pretty safe, but I suppose there could come in a change during fetch from another part using stream I suppose..

will make it more fail safe

llucax commented on src/frequenz/dispatch/_bg_service.py:
```diff
@@ -396,23 +396,30 @@ async def _fetch(self, timer: Timer) -> None:
 
-        for dispatch in old_dispatches.values():
+        # Delete old dispatches
+        for dispatch_id in old_dispatches:
+            if dispatch_id in new_dispatches:
+                continue
```
You could probably also do something like:
```suggestion
        for dispatch_id in old_dispatches.keys() - new_dispatches.keys():
```

llucax reviewed (dismissed):
LGTM, but one thing to check and one small suggestion.

Marenz commented on src/frequenz/dispatch/_bg_service.py:
```diff
@@ -381,9 +381,9 @@ async def _fetch(self, timer: Timer) -> None:
-                    old_dispatch = old_dispatches.pop(dispatch.id, None)
-                    if not old_dispatch:
+                    new_dispatches[dispatch.id] = dispatch
+                    old_dispatch = self._dispatches.get(dispatch.id, None)
```
It's nice for the reader to see this explicitly I think

llucax commented on src/frequenz/dispatch/_bg_service.py:
```diff
@@ -366,8 +366,8 @@ async def _fetch(self, timer: Timer) -> None:
 
-        old_dispatches = self._dispatches
-        self._dispatches = {}
+        old_dispatches = set(self._dispatches.keys())
```
```suggestion
```

llucax commented on src/frequenz/dispatch/_bg_service.py:
```diff
@@ -396,23 +396,33 @@ async def _fetch(self, timer: Timer) -> None:
-            await self._lifecycle_events_tx.send(Deleted(dispatch=dispatch))
-            await self._update_dispatch_schedule_and_notify(None, dispatch, timer)
+        # Delete old dispatches
+        for dispatch_id in old_dispatches - new_dispatches.keys():
+            try:
+                dispatch = self._dispatches.pop(dispatch_id)
+                _logger.debug("Deleted dispatch: %s", dispatch)
+                await self._update_dispatch_schedule_and_notify(None, dispatch, timer)
+
+                # Set deleted only here as it influences the result of dispatch.started
+                # which is used in the func call above ^
+                dispatch._set_deleted()  # pylint: disable=protected-access
+                await self._lifecycle_events_tx.send(Deleted(dispatch=dispatch))
+            except KeyError:
+                _logger.warning(
+                    "Inconsistency in cache detected. Tried to delete non-existing dispatch %s",
+                    dispatch_id,
+                )
```
It is usually recommended to narrow the try/except to only the code you expect to fail, otherwise you could catch other failures that maybe are not OK to ignore. For example if some code in `await self._lifecycle_events_tx.send(Deleted(dispatch=dispatch))` (suppose `send()` uses some internal cache and there is a bug that also make it raise a `KeyError`).

Also if `old_dispatches` == `self._dispatches.keys()`, maybe it makes more sense to use `self._dispatches.keys()` directly now, I see in the new code `old_dispatches` is only being used here. This would make the code more obvious and actually guarantee the `pop()` can't fail, `old_dispatches` and `self._dispatches` can't get out of sync.

So I suggest removing `old_dispatches` and just do:

```suggestion
        # We make a copy because we mutate self._dispatch.keys() inside the loop
        for dispatch_id in frozenset(self._dispatches.keys() - new_dispatches.keys()):
            dispatch = self._dispatches.pop(dispatch_id)
            _logger.debug("Deleted dispatch: %s", dispatch)
            await self._update_dispatch_schedule_and_notify(None, dispatch, timer)

            # Set deleted only here as it influences the result of dispatch.started
            # which is used in the func call above ^
            dispatch._set_deleted()  # pylint: disable=protected-access
                await self._lifecycle_events_tx.send(Deleted(dispatch=dispatch))
```

llucax replied:
Not sure if the copy is necessary, as we are using the `-` which I guess produces a copy in itself, but I'm not 100% sure. Also making it a `frozenset` guarantees our items being iterated will never change, so it makes everything more explicit.

Marenz replied:
> and actually guarantee the pop() can't fail, 

I don't think that's actually true. It can still fail, the maps can be modified between any call to await.

llucax replied:
Oh, OK ,yeah, then maybe it is worth a comment in the `try`, as it might not be obvious to the casual reader.

llucax reviewed (commented):
LGTM, just one more suggestion, removing auto-merge in case you want to change it.

Marenz commented on src/frequenz/dispatch/_merge_strategies.py:
```diff
@@ -39,24 +42,38 @@ def filter(
-        logging.debug(
-            "stop event %s because other_dispatches_running=%s",
+        other_dispatches_running = any(
+            dispatch.started_at(now) for dispatch in running_dispatch_list
```
Slightly surprised pylint or flake didn't catch that

Marenz commented on src/frequenz/dispatch/_bg_service.py:
```diff
@@ -396,23 +395,38 @@ async def _fetch(self, timer: Timer) -> None:
-        for dispatch in old_dispatches.values():
+        # We make a copy because we mutate self._dispatches.keys() inside the loop
+        for dispatch_id in frozenset(self._dispatches.keys() - new_dispatches.keys()):
+            try:
+                dispatch = self._dispatches.pop(dispatch_id)
+            except KeyError as error:
+                _logger.warning(
+                    "Inconsistency in cache detected. "
+                    + "Tried to delete non-existing dispatch %s (%s)",
+                    dispatch_id,
+                    error,
+                )
+                continue
+
```
shush

llucax reviewed (dismissed):
LGTM too, one small optional comment about maybe adding a comment to the `try` for `pop()` as it might not be obvious.
